plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,553 @@
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks_payload->len);
+
+	for (guint i = 0; i < chunks_payload->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks_payload, i);
+		g_autoptr(GByteArray) part_firmware = g_byte_array_new();
+
+		g_byte_array_append(part_firmware,
+				    fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk));
+
+		fu_byte_array_set_size(transfer_parameter, 4, 0);
+		if ((i + 1) % 4 == 0)
+			transfer_parameter->data[0] = 1;
+		else
+			transfer_parameter->data[0] = 0;
+
+		fu_memwrite_uint24(transfer_parameter->data + 1,
+				   fu_chunk_get_address(chk),
+				   G_BIG_ENDIAN);
+
+		value = fu_memread_uint16(transfer_parameter->data, G_BIG_ENDIAN);
+		index = fu_memread_uint16(transfer_parameter->data + 2, G_BIG_ENDIAN);
+
+		if (!algoltek_device_ctrl_transfer(self,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   FU_ALGOLTEK_CMD_WRF,
+						   value,
+						   index,
+						   part_firmware,
+						   part_firmware->len,
+						   error))
+			return FALSE;
+
+		if ((i + 1) % 4 == 0 || (i + 1) == chunks_payload->len) {
+			if (!fu_device_retry_full(FU_DEVICE(self),
+						  fu_algoltek_device_status_check,
+						  10,
+						  0,
+						  NULL,
+						  error))
+				return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) version_data = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = algoltek_device_rdv(self, error);
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	fu_device_set_version(device, version_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autoptr(GBytes) blob_isp = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 18, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 2, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, NULL);
+
+	if (!algoltek_device_en(self, error)) {
+		g_prefix_error(error, "system activation failure: ");
+		return FALSE;
+	}
+
+	if (!algoltek_device_rst(self, 0x200, error)) {
+		g_prefix_error(error, "system reboot failure: ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	if (!algoltek_device_wrr(self, 0x80AD, 0, error)) {
+		g_prefix_error(error, "data write failure: ");
+		return FALSE;
+	}
+
+	if (!algoltek_device_wrr(self, 0x80C0, 0, error))
```

Comment:
Can all these wrr calls be part of a for loop instead? It seems like the argument counts up predicably.

Reply:
No, there are hardware settings defined by firmware. These addresses may vary and it's not periodical.

Reply:
Thanks!